### PR TITLE
Add clickhouse support in metrics

### PR DESCRIPTION
**Pull request by Liuhaai**
add additional clickhouse component in metrics center

### CODE REVIEW

File: pkg/modules/metrics/clickhouse.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	sleepTime     = 10 * time.Second
+)
+
+func Init(ctx context.Context) {
+	cfg, existed := types.MetricsCenterConfigFromContext(ctx)
+	if !existed ||
+		len(cfg.ClickHouseAddr) == 0 ||
+		len(cfg.ClickHouseDB) == 0 ||
+		len(cfg.ClickHouseUser) == 0 ||
+		len(cfg.ClickHousePassword) == 0 {
```
Comment (hunshenshi) [inline pkg/modules/metrics/clickhouse.go:41]: impl `IsZero()` func

File: pkg/modules/metrics/clickhouse.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+				log.Println("ClickhouseClient failed to connect: ", err)
+				time.Sleep(sleepTime)
+				continue
+			}
+		}
+		query := <-c.sqLQueue
+		if err := c.conn.AsyncInsert(context.Background(), query, true); err != nil {
+			if !c.liveness() {
+				c.conn = nil
+				c.sqLQueue <- query
```
Comment (hunshenshi) [inline pkg/modules/metrics/clickhouse.go:83]: may blocked

File: pkg/modules/metrics/clickhouse.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	c.sqLQueue <- query
+	return nil
+}
+
+func (c *ClickhouseClient) Run() {
+	for {
+		if c.conn == nil {
+			if err := c.connect(); err != nil {
+				log.Println("ClickhouseClient failed to connect: ", err)
+				time.Sleep(sleepTime)
```
Comment (hunshenshi) [inline pkg/modules/metrics/clickhouse.go:75]: should use a incremental step?

**Review by hunshenshi — APPROVED** [review]

File: pkg/modules/metrics/clickhouse.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+		cfg.ClickHouseUser,
+		cfg.ClickHousePassword,
+	})
+}
+
+func validateCfg(cfg *types.MetricsCenterConfig) bool {
+	return len(cfg.ClickHouseAddr) > 0 &&
+		len(cfg.ClickHouseDB) > 0 &&
+		len(cfg.ClickHouseUser) > 0 &&
+		len(cfg.ClickHousePassword) > 0
```
Comment (CoderZhi) [inline pkg/modules/metrics/clickhouse.go:53]: decode, validate and return parameters

File: pkg/modules/metrics/clickhouse.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+func newClickhouseClient(cfg *config) *ClickhouseClient {
+	cc := &ClickhouseClient{
+		sqLQueue: make(chan string),
```
Comment (CoderZhi) [inline pkg/modules/metrics/clickhouse.go:58]: specify channel length here?

File: pkg/modules/metrics/clickhouse.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	go cc.Run()
+	return cc
+}
+
+func (c *ClickhouseClient) Insert(query string) error {
+	if len(c.sqLQueue) > queueLength {
+		return errors.New("the queue of client is full")
+	}
+	c.sqLQueue <- query
+	return nil
```
Comment (CoderZhi) [inline pkg/modules/metrics/clickhouse.go:70]: any lock?

File: pkg/modules/metrics/clickhouse.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+		if c.conn == nil {
+			if err := c.connect(); err != nil {
+				log.Println("ClickhouseClient failed to connect: ", err)
+				time.Sleep(sleepTime)
+				continue
+			}
+		}
+		query := <-c.sqLQueue
+		if err := c.conn.AsyncInsert(context.Background(), query, true); err != nil {
+			if !c.liveness() {
```
Comment (CoderZhi) [inline pkg/modules/metrics/clickhouse.go:84]: error may not be caused by liveness, and may cause `liveness() == false`. In that case, we may fail to catch the root cause of the error.

File: pkg/modules/metrics/clickhouse.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+				c.conn = nil
+				c.sqLQueue <- query
+				continue
+			}
+			log.Println("ClickhouseClient failed to insert data: ", err)
```
Comment (CoderZhi) [inline pkg/modules/metrics/clickhouse.go:89]: if liveness == true, reinsert query.

File: pkg/modules/metrics/clickhouse.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+				log.Println("ClickhouseClient failed to connect: ", err)
+				time.Sleep(sleepTime)
+				continue
+			}
+		}
+		query := <-c.sqLQueue
+		if err := c.conn.AsyncInsert(context.Background(), query, true); err != nil {
+			if !c.liveness() {
+				c.conn = nil
+				c.sqLQueue <- query
```
Comment (CoderZhi) [inline pkg/modules/metrics/clickhouse.go:86]: does the order matter? channel is strictly FIFO

File: cmd/srv-applet-mgr/config/config.yml.template [hunk 10]
```diff
@@ -18,6 +18,10 @@ SRV_APPLET_MGR__Logger_Level: debug
 SRV_APPLET_MGR__Logger_Name: srv-applet-mgr
 SRV_APPLET_MGR__Logger_Output: ALWAYS
 SRV_APPLET_MGR__Logger_ReportCaller: "false"
+SRV_APPLET_MGR__MetricsCenter_ClickHouseAddr: ""
+SRV_APPLET_MGR__MetricsCenter_ClickHouseDB: ""
+SRV_APPLET_MGR__MetricsCenter_ClickHousePassword: ""
+SRV_APPLET_MGR__MetricsCenter_ClickHouseUser: ""
```
[the remark this answers is not in the document]
Comment (Liuhaai) [reply cmd/srv-applet-mgr/config/config.yml.template:24]: any idea for the format?

File: pkg/modules/metrics/clickhouse.go (hunk 6 above)
Comment (Liuhaai) [reply pkg/modules/metrics/clickhouse.go:70]: yes, the queue length isn't accurate here

File: cmd/srv-applet-mgr/config/config.yml.template (hunk 10 above)
[the remark this answers is not in the document]
Comment (Liuhaai) [reply cmd/srv-applet-mgr/config/config.yml.template:24]: https://github.com/ClickHouse/ClickHouse/issues/29880

**Pull request merged**